library/std/src/diagnostics.qs

```diff
@@ -69,6 +69,44 @@ namespace Microsoft.Quantum.Diagnostics {
         body intrinsic;
     }
 
+    /// # Summary
+    /// Given an operation, dumps the matrix representation of the operation actiong on the given
+    /// number of qubits.
+    ///
+    /// # Input
+    /// ## nQubits
+    /// The number of qubits on which the given operation acts.
+    /// ## op
+    /// The operation that is to be diagnosed.
+    ///
+    /// # Remarks
+    /// When run on the sparse-state simulator, the following snippet
+    /// will output the matrix
+    /// $\left(\begin{matrix} 0.0 & 0.707 \\\\ 0.707 & 0.0\end{matrix}\right)$:
+    ///
+    /// ```qsharp
+    /// operation DumpH() : Unit {
+    ///     DumpOperation(1, qs => H(qs[0]));
+    /// }
+    /// ```
+    /// Calling this operation has no observable effect from within Q#.
+    /// Note that if `DumpOperation` is called when there are other qubits allocated,
+    /// the matrix displayed may reflect any global phase that has accumulated on the other qubits.
+    operation DumpOperation(nQubits : Int, op : Qubit[] => Unit) : Unit {
+        use (targets, extra) = (Qubit[nQubits], Qubit[nQubits]);
```

Comment:
This is a tricky one. My first instinct was to say that this code shouldn't be called if the user is doing QIR codegen, but there isn't anything in here that shows them directly that it should be commented out. But I think you are on to something with using `SimulatableIntrinsic` to avoid this. Let me try a few things and see if the behavior matches expectation.

In general though, `DumpOperation` can have side effects even during simulation that make it undesirable to include in code that is being used for codegen, since it won't always behave the way the user expects.

Reply:
Ok, your suggestion did just the trick! I made `DumpOperation` and simulatable intrinsic and added it to the list of recognized no-op callables in partial evaluation. That is enough to preserve its behavior in simulation but cause it to be fully skipped (like `DumpMachine` and `DumpRegister`) during QIR codegen.